### ISSUE TITLE
[Support] Are tables typed, or just the individual elements?

### ISSUE THREAD

**Issue by 7ombie** [WebAssembly/design#1412]
Just a quick question... Is the intention to allow multiple tables, where each is restricted to a specific type of reference (either `funcref` or `externref`), or can each table hold any combination of the two reference types?

**Comment by tlively** [comment]
All slots in a table must contain values of the table's type, so `funcref` and `externref` values cannot be placed in the same table unless the table's type is `anyref`.

**Comment by 7ombie** [comment]
Ahhh, that makes sense. Thank you, @tlively. Much appreciated.

**Comment by 7ombie** [comment]
@tlively - Sorry to be a pain. Do you know how to encode a table's type as `anyref` (the byte used in the binary format). I can only find the encodings for `funcref` and `externref`.

**Comment by tlively** [comment]
`anyref` is introduced in the GC proposal and its encoding is -0x12: https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#reference-types-1.

**Comment by 7ombie** [comment]
Thanks, @tlively. It wasn't obvious that, for example, signed `-0x11` encodes to the the same byte as unsigned `0x6F`.

The following note from the spec is confusing:

> Value types can occur in contexts where type indices are also allowed, such as in the case of block types. Thus, the binary format for types corresponds to the signed LEB128 encoding of small negative s𝑁
 values, so that they can coexist with (positive) type indices in the future.

In plain English, does this mean that valtypes use a signed LEB128 encoding (for future proofing)?

**Comment by 7ombie** [comment]
Is there a place online where people like myself (who are still learning WebAssembly) can ask these kinds of questions? I don't want to spam the issue tracker, if there's a better place to post this kind of thing.

**Comment by ngzhian** [comment]
> Is there a place online where people like myself (who are still learning WebAssembly) can ask these kinds of questions? I don't want to spam the issue tracker, if there's a better place to post this kind of thing.

You might be interested in the WebAssembly discord channel, see invite link here https://webassembly.org/community/resources/.

**Comment by 7ombie** [comment]
Thank you, @ngzhian. Much appreciated.

**Comment by fgmccabe** [comment]
How can -0x11 encode to 0x6c? This should not be possible.

**Comment by tlively** [comment]
> > Value types can occur in contexts where type indices are also allowed, such as in the case of block types. Thus, the binary format for types corresponds to the signed LEB128 encoding of small negative s𝑁
> > values, so that they can coexist with (positive) type indices in the future.
> 
> In plain English, does this mean that valtypes are now (effectively) decoded as signed LEB128 (for future proofing), so the old unsigned encodings have been remapped to the signed equivalents (for backwards compatibility)? If so, why does the official spec still specify the positive encodings for types?

Yes, the numbers for the types were chosen to be future-compatible with a spec change to interpret them as signed LEBs. As for why the spec didn't just treat them as signed LEBs all along, I'm not sure there's a great reason. It's more conservative to just treat them as single bytes rather than allowing a variable-length encoding, but it would not have caused any issues to treat them as LEBs from the start. At this point it's probably not worth changing until the positive numbers are actually going to be used for something.
